internal-services/catalog/iib-add-fbc-fragment-to-index-image-task.yaml

```diff
@@ -137,57 +136,89 @@ spec:
             fbcOptIn="false"
             mustSignIndexImage="false"
             mustPublishIndexImage="false"
+        else
+            # check if this fbc fragment is opt-in
+            echo "Fetching the image bundle from $(params.fbcFragment)..."
+            PULL_SPEC_LIST=$(opm render "$(params.fbcFragment)" | jq -r \
+            'select(.schema == "olm.bundle") | "\(.image)" | split("@")[0]' |uniq)
+            for PULL_SPEC in ${PULL_SPEC_LIST}; do
+                # make sure they query is done using the internal name instead of the public
+                PULL_SPEC="${PULL_SPEC//registry.redhat.io/registry.access.redhat.com}"
 
-        # in case of more than one image all should have the same fbc_opt_in value
-        elif [ $(wc -w <<< ${fbcOptIn[@]}) -ne 1 ]; then
-            fbcOptIn="false"
-            mustSignIndexImage="false"
-            mustPublishIndexImage="false"
+                echo "Attempting to fetch from ${FETCH_URL} to check if fragment is \`fbc_opt_in==true\`..."
+                IFS=" " read -r -a fbcOptIn <<< "$(isfbcOptIn "${PULL_SPEC}")"
+            done
+
+            # fbcOptIn=($(printf "%s\n" "${fbcOptIn[@]}" |sort |uniq))
+            IFS=" " read -r -a fbcOptIn <<< "$(printf "%s\n" "${fbcOptIn[@]}" |sort |uniq)"
+            mustPublishIndexImage="${fbcOptIn[0]}"
+            mustSignIndexImage="${fbcOptIn[0]}"
+
+            # in case of more than one image all should have the same fbc_opt_in value
+            if [ "$(wc -w <<< "${fbcOptIn[@]}")" -ne 1 ]; then
+                fbcOptIn=("false")
+                mustSignIndexImage="false"
+                mustPublishIndexImage="false"
+            fi
         fi
 
-        echo "Fragment has \`fbc_opt_in==${fbcOptIn}\`"
+        echo "Fragment has \`fbc_opt_in==${fbcOptIn[0]}\`"
```

Comment:
This is not right. `fbcOptIn` is simply set to `"false"` for hotfix and for stage builds. But then an array for the last branch of the `if`. I would suggest to just use a string.

In the `else` branch, where you check each image separately, what you're checking is that all of them need to have `fbcOptIn=true`, right? If at least one of them has false, then you cannot publish it. So before the loop, set `fbcOptIn=true`. And in each iteration, run `isfbcOptIn` and if it's false, stop (`break`) right there - you know for sure that you cannot publish it.

Reply:
this is one of the reasons I wanted initially to add the shellcheck exclusions. I would prefer rewrite the full fbc_opt_in part because this is looking ugly. I really can't remember why eventually it would return more than 1 items in the array, and I was a bit afraid to break the builds by trying to fix this now.

Anyhow the change done is basically trusting shellcheck that informed that the above code would return the first element of the array in all cases, so that why I pinned the 0, so it passes (in theory) without change the behavior.

What would you suggest? Should it fix it properly in this ticket or comment out this and create a fix ticket for it? Both works for me.

Reply:
Does this relate to making the task idempotent? I am getting confused

Reply:
OK, if you only touched this part to make shell check happy, I'm fine adding shellcheck ignores (but only for the affected lines, not at the top) and creating a follow up jira to fix it later.

But it definitely is not right to treat it as an array here when we can clearly see that in some cases it will just be a string.

Reply:
Do we need a follow up ticket? Is the existing one about fixing shellcheck issues in this task not enough?

Reply:
can be as well. I will add some more info to the current ticket then.  Because the logic needs  a bit of change, not only the syntax  check.